sensors/sensors.c

```diff
@@ -0,0 +1,630 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Sensor Manager
+ *
+ * Copyright 2022 Phoenix Systems
+ * Author: Hubert Buczynski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <errno.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <string.h>
+#include <sys/msg.h>
+#include <sys/stat.h>
+#include <sys/list.h>
+#include <sys/threads.h>
+#include <posix/utils.h>
+
+#include "sensors.h"
+
+#define CLIENT_SET_ID(id) (id + 1)
+#define CLIENT_GET_ID(id) (id - 1)
+
+#define NB_SENSOR_TYPES (sizeof(sensor_type_t) * 8)
+
+
+typedef struct {
+	sensors_ops_t ops;
+	int refs;
+	idnode_t node;
+} sensor_client_t;
+
+
+struct {
+	rbtree_t drvs;    /* registered sensors drivers */
+	idtree_t infos;   /* driver instances */
+	idtree_t clients; /* set of clients */
+
+	handle_t cLock;                          /* client's tree lock */
+	handle_t locks[NB_SENSOR_TYPES];         /* locks for each sensor type */
+	uint8_t evtNb[NB_SENSOR_TYPES];          /* number of events from all sensors */
+	sensor_event_t *events[NB_SENSOR_TYPES]; /* each row defines set of events of the same type */
+
+	uint8_t **devEvents; /* events assign to each device */
+
+	oid_t oid;
+} sensors_common;
+
+
+/* Update data from sensors */
+
+int sensors_publish(unsigned int devId, const sensor_event_t *event)
+{
+	int err = EOK;
+	uint8_t id = __builtin_ffs(event->type);
+	uint8_t evtId = sensors_common.devEvents[devId][id];
```

Comment:
This is a bad access - id hasn't been decremented yet, it's outside of a mutex

Reply:
I agree that there is a bad access. However, the `sensors_common.devEvent` is read only, so that it doesn't have to be protected by the mutex.